src/java/htsjdk/samtools/CRAMFileReader.java

```diff
@@ -562,64 +495,46 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
 
         @Override
         public void close() {
-            delegate.close();
+            if (unfilteredIterator != null) {
+                unfilteredIterator.close();
+            }
         }
 
         @Override
         public boolean hasNext() {
-            if (next != null) return true;
-            if (noMore) return false;
-
-            while (delegate.hasNext()) {
-                next = delegate.next();
-
-                if (isWithinTheInterval(next)) break;
-                if (isBeyondTheInterval(next)) {
-                    next = null;
-                    noMore = true;
-                    return false;
-                }
-                next = null;
+            if (nextRec != null) {
+                return true;
+            } else if (!hasMore) {
+                return false;
             }
-
-            return next != null;
-        }
-
-        boolean isWithinTheInterval(final SAMRecord record) {
-            final boolean refMatch = record.getReferenceIndex() == interval.referenceIndex;
-            if (interval.start == -1) return refMatch;
-            if (!refMatch) return false;
-
-            final int start = record.getAlignmentStart();
-            final int end = record.getAlignmentEnd();
-            if (overlap) {
-                return CoordMath.overlaps(start, end, interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end);
-            } else {
-                // contained:
-                return CoordMath.encloses(interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end, start, end);
+            while (nextRec == null && hasMore && unfilteredIterator.hasNext()) {
```

Comment:
Why do you need both `hasMore` and `nextRec`? Can you just use `nextRec == null` to signal the end of the iteration (this is what most iterators do)?